torch/optim/asgd.py

```diff
@@ -76,19 +80,117 @@ def step(self, closure=None):
-                   lambd=group['lambd'])
+            asgd(params_with_grad,
+                 grads,
+                 axs,
+                 mus,
+                 etas,
+                 weight_decay=group['weight_decay'],
+                 lambd=group['lambd'],
+                 foreach=group['foreach'])
 
             # update eta and mu
-            for p, mu, eta in zip(params_with_grad, mus, etas):
+            for p in params_with_grad:
                 state = self.state[p]
                 state['eta'] = (group['lr'] /
                                 math.pow((1 + group['lambd'] * group['lr'] * state['step']), group['alpha']))
                 state['mu'] = 1 / max(1, state['step'] - group['t0'])
 
         return loss
+
+
+def asgd(params: List[Tensor],
+         grads: List[Tensor],
+         axs: List[Tensor],
+         mus: List[float],
+         etas: List[float],
+         foreach: bool = None,
+         *,
+         weight_decay: float,
+         lambd: float):
+    r"""Functional API that performs asgd algorithm computation.
+
+    See :class:`~torch.optim.ASGD` for details.
+    """
+
+    if foreach is None:
+        # Placeholder for more complex foreach logic to be added when value is not set
+        foreach = False
+
+    if foreach and torch.jit.is_scripting():
+        raise RuntimeError('torch.jit.script not supported with foreach optimizers')
+
+    if foreach and not torch.jit.is_scripting():
+        func = _multi_tensor_asgd
+    else:
+        func = _single_tensor_asgd
+
+    func(params,
+         grads,
+         axs,
+         mus,
+         etas,
+         weight_decay=weight_decay,
+         lambd=lambd)
+
+
+def _single_tensor_asgd(params: List[Tensor],
+                        grads: List[Tensor],
+                        axs: List[Tensor],
+                        mus: List[float],
+                        etas: List[float],
+                        *,
+                        weight_decay: float,
+                        lambd: float):
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        mu = mus[i]
+        ax = axs[i]
+        eta = etas[i]
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # decay term
+        param.mul_(1 - lambd * eta)
+
+        # update parameter
+        param.add_(grad, alpha=-eta)
+
+        # averaging
+        if mu != 1:
+            ax.add_(param.sub(ax).mul(mu))
+        else:
+            ax.copy_(param)
+
+
+def _multi_tensor_asgd(params: List[Tensor],
+                       grads: List[Tensor],
+                       axs: List[Tensor],
+                       mus: List[float],
+                       etas: List[float],
+                       *,
+                       weight_decay: float,
+                       lambd: float):
+
+    if len(params) == 0:
+        return
+
+    if weight_decay != 0:
+        torch._foreach_add_(grads, params, alpha=weight_decay)
+
+    # decay term
```

Comment:
The initial `multitensor/_functional` ASGD logic assumes that eta is the same for all params passed to the function, (`eta = states[0]['eta']` where states[0] is the state for the first param) and then (`torch._foreach_mul_(params, 1 - lambd * eta)`) I wasn't sure whether this should be the case as it seemed slightly different from the other optimizers, so I made a change here, let me know if i should change it back!!

Reply:
Again I think it is safer to not change the behavior in this PR.
We could add a check that `all(states[0]['eta'] == states[i]['eta'] for i in range(len(states)))` if you think this is a bug. But I would prefer delay any change of behavior to a later PR to keep things simple in case we break things and we have to track back and revert some of these changes.